plugins/uast-kotlin-fir/src/org/jetbrains/uast/kotlin/FirKotlinUastLanguagePlugin.kt

```diff
@@ -0,0 +1,68 @@
+/*
+ * Copyright 2010-2021 JetBrains s.r.o. and Kotlin Programming Language contributors.
+ * Use of this source code is governed by the Apache 2.0 license that can be found in the license/LICENSE.txt file.
+ */
+
+package org.jetbrains.uast.kotlin
+
+import com.intellij.lang.Language
+import com.intellij.psi.PsiElement
+import org.jetbrains.kotlin.idea.KotlinLanguage
+import org.jetbrains.uast.DEFAULT_TYPES_LIST
+import org.jetbrains.uast.UElement
+import org.jetbrains.uast.UExpression
+import org.jetbrains.uast.UastLanguagePlugin
+import org.jetbrains.uast.kotlin.FirKotlinConverter.convertDeclarationOrElement
+
+interface FirKotlinUastService {
```

Comment:
One more quick thought.

The counterpart, `KotlinUastResolveProviderService`, has `getBindingContext` API, through which CLI v.s. IDE modes can supply binding context differently.

Maybe, we can do something similar here, like passing different FIR provider: CLI will use `FirAnalyzerFacade` directly, while IDE will use `KtAnalysisSession` and so on. Then, current FIR element retrievals need to be rewritten to use such provider.

Hm, so `...ResolveProvider...` makes sense, perhaps this needs to be renamed as well, e.g., `FirKotlinUastResolveProviderService`

Reply:
We can pass a different KtAnalysisSession factories. One will be lazy as in IDE, the other one will be non-lazy

Reply:
I wonder we don't have non-lazy version yet? In any case, it'd be appreciated if you can let me know some pointers to KtAnalysisSession factories and use cases.

Reply:
There is no non-lazy version now, but it should be not very hard to implement one by creating a non-lazy implementation `FirModuleResolveState` which holds all resolve inside analysis session. We are currently reworking the `FirModuleResolveState`, after rework is done we may think about non-lazy version of it.

KtAnalysisSessionProvider may be used for providing needed KtAnalysisSession.

Reply:
> Just want to confirm my understanding, so we will make `KtAnalysisSession` works in CLI mode as well, correct? Then we will define a set of API on `KtAnalysisSession` that operates on FIR based `UElement`, just like the `KtSymbol` that we already have.

`UElement` generated in this module will use `KtAnalysisSession` (per mode) to use `KtSymbol`. So, at this moment, I'm not sure what APIs we will need.

Reply:
We will be able to use all API which is present in lazy KtAnalysysSession, all its takes to properly  implement non-lazy version of FirModuleResolveState as FirModuleResolveState is responsible for lazy-resolving